File: Source/CPP/PgSqlBackend.cpp
#include "PgSqlBackend.hpp"
#include "PgSqlResult.hpp"
#include "../../../LangShared/InternalString/CPP/InternalString.hpp"
#include <string>
#include <vector>

namespace NumbatLogic
{
	static void SplitParamString(const char* sxParamString, int nParams, std::vector<std::string>& out)
	{
		out.clear();
		if (!sxParamString || nParams <= 0)
			return;
		out.reserve(static_cast<size_t>(nParams));
		const char* p = sxParamString;
		for (int i = 0; i < nParams; i++)
		{
			const char* start = p;
			while (*p != '\0' && *p != '\x01')
				p++;
			out.push_back(std::string(start, static_cast<size_t>(p - start)));
			if (*p == '\x01')
				p++;
		}
	}
	static void SetLastError(InternalString*& sErr, const char* szMsg)
	{
		if (sErr)
			delete sErr;
		sErr = new InternalString(szMsg ? szMsg : "");
	}

	static void ClearLastError(InternalString*& sErr)
	{
		if (sErr)
		{
			delete sErr;
			sErr = nullptr;
		}
	}

	PgSqlBackend::PgSqlBackend()
		: m_pConn(nullptr)
		, m_sLastError(nullptr)
	{
	}

	PgSqlBackend::~PgSqlBackend()
	{
		if (m_pConn)
		{
			PQfinish(m_pConn);
			m_pConn = nullptr;
		}
		delete m_sLastError;
		m_sLastError = nullptr;
	}

	bool PgSqlBackend::Connect(const char* sxHost, const char* sxUsername, const char* sxPassword, const char* sxDatabase)
	{
		if (m_pConn)
		{
			PQfinish(m_pConn);
			m_pConn = nullptr;
		}

		char connInfo[1024];
		snprintf(connInfo, sizeof(connInfo),
			"host=%s user=%s password=%s dbname=%s",
			sxHost ? sxHost : "",
			sxUsername ? sxUsername : "",
			sxPassword ? sxPassword : "",
			sxDatabase ? sxDatabase : "");

		m_pConn = PQconnectdb(connInfo);

		if (PQstatus(m_pConn) != CONNECTION_OK)
		{
			SetLastError(m_sLastError, PQerrorMessage(m_pConn));
			PQfinish(m_pConn);
			m_pConn = nullptr;
			return false;
		}
		ClearLastError(m_sLastError);
		return true;
	}

	bool PgSqlBackend::Exec(const char* sxSql)
	{
		if (!m_pConn || !sxSql)
		{
			ClearLastError(m_sLastError);
			return false;
		}

		PGresult* pResult = PQexec(m_pConn, sxSql);

		if (!pResult)
		{
			SetLastError(m_sLastError, PQerrorMessage(m_pConn));
			return false;
		}

		ExecStatusType status = PQresultStatus(pResult);
		if (status != PGRES_COMMAND_OK && status != PGRES_TUPLES_OK && status != PGRES_EMPTY_QUERY)
		{
			SetLastError(m_sLastError, PQresultErrorMessage(pResult));
			PQclear(pResult);
			return false;
		}
		PQclear(pResult);
		ClearLastError(m_sLastError);
		return true;
	}

	bool PgSqlBackend::QueryExists(const char* sxSql)
	{
		if (!m_pConn || !sxSql)
		{
			ClearLastError(m_sLastError);
			return false;
		}

		PGresult* pResult = PQexec(m_pConn, sxSql);

		if (!pResult)
		{
			SetLastError(m_sLastError, PQerrorMessage(m_pConn));
			return false;
		}

		ExecStatusType status = PQresultStatus(pResult);
		bool bHasRows = (status == PGRES_TUPLES_OK && PQntuples(pResult) > 0);
		if (!bHasRows && status != PGRES_TUPLES_OK && status != PGRES_EMPTY_QUERY)
			SetLastError(m_sLastError, PQresultErrorMessage(pResult));
		else
			ClearLastError(m_sLastError);
		PQclear(pResult);
		return bHasRows;
	}

	const char* PgSqlBackend::GetLastError()
	{
		return m_sLastError ? m_sLastError->GetExternalString() : "";
	}

	InternalString* PgSqlBackend::QueryFirstRowSingleColumn(const char* sxSql)
	{
		if (!m_pConn || !sxSql)
		{
			ClearLastError(m_sLastError);
			return nullptr;
		}
		PGresult* pResult = PQexec(m_pConn, sxSql);
		if (!pResult)
		{
			SetLastError(m_sLastError, PQerrorMessage(m_pConn));
			return nullptr;
		}
		if (PQresultStatus(pResult) != PGRES_TUPLES_OK || PQntuples(pResult) < 1 || PQnfields(pResult) < 1)
		{
			PQclear(pResult);
			return nullptr;
		}
		const char* v = PQgetvalue(pResult, 0, 0);
		InternalString* pOut = new InternalString(v ? v : "");
		PQclear(pResult);
		ClearLastError(m_sLastError);
		return pOut;
	}

	PgSqlResult* PgSqlBackend::ExecuteQuery(const char* sxSql, int nParams, const char* sxParamString)
	{
		if (!m_pConn || !sxSql)
		{
			ClearLastError(m_sLastError);
			return nullptr;
		}
		std::vector<std::string> paramStorage;
		SplitParamString(sxParamString, nParams, paramStorage);
		std::vector<const char*> paramPtrs;
		for (size_t i = 0; i < paramStorage.size(); i++)
			paramPtrs.push_back(paramStorage[i].c_str());
		const char* const* paramValues = paramPtrs.empty() ? nullptr : paramPtrs.data();
		PGresult* pResult = PQexecParams(m_pConn, sxSql, nParams, nullptr, paramValues, nullptr, nullptr, 0);
		if (!pResult)
		{
			SetLastError(m_sLastError, PQerrorMessage(m_pConn));
			return nullptr;
		}

		ExecStatusType eStatus = PQresultStatus(pResult);
		if (eStatus != PGRES_TUPLES_OK && eStatus != PGRES_COMMAND_OK)
		{
			SetLastError(m_sLastError, PQresultErrorMessage(pResult));
			PQclear(pResult);
			return nullptr;
		}

		ClearLastError(m_sLastError);
		return new PgSqlResult(pResult);
	}
}

File: Source/CS/PgSqlBackend.cs
using System;
using System.Collections.Generic;
using Npgsql;

namespace NumbatLogic
{
	public class PgSqlBackend
	{
		private NpgsqlConnection m_pConn;
		private string m_sLastError;

		public PgSqlBackend()
		{
			m_pConn = null;
			m_sLastError = null;
		}

		private void SetLastError(string szMsg)
		{
			m_sLastError = szMsg ?? string.Empty;
		}

		private void ClearLastError()
		{
			m_sLastError = null;
		}

		public bool Connect(string sxHost, string sxUsername, string sxPassword, string sxDatabase)
		{
			try
			{
				if (m_pConn != null)
				{
					m_pConn.Dispose();
					m_pConn = null;
				}

				var builder = new NpgsqlConnectionStringBuilder
				{
					Host = sxHost ?? string.Empty,
					Username = sxUsername ?? string.Empty,
					Password = sxPassword ?? string.Empty,
					Database = sxDatabase ?? string.Empty
				};

				m_pConn = new NpgsqlConnection(builder.ConnectionString);
				m_pConn.Open();
				ClearLastError();
				return true;
			}
			catch (Exception ex)
			{
				SetLastError(ex.Message);
				if (m_pConn != null)
				{
					m_pConn.Dispose();
					m_pConn = null;
				}
				return false;
			}
		}

		public bool Exec(string sxSql)
		{
			if (m_pConn == null || sxSql == null)
			{
				ClearLastError();
				return false;
			}

			try
			{
				using (var cmd = new NpgsqlCommand(sxSql, m_pConn))
				{
					cmd.ExecuteNonQuery();
				}
				ClearLastError();
				return true;
			}
			catch (Exception ex)
			{
				SetLastError(ex.Message);
				return false;
			}
		}

		public bool QueryExists(string sxSql)
		{
			if (m_pConn == null || sxSql == null)
			{
				ClearLastError();
				return false;
			}

			try
			{
				using (var cmd = new NpgsqlCommand(sxSql, m_pConn))
				using (var reader = cmd.ExecuteReader())
				{
					bool hasRows = reader.HasRows;
					ClearLastError();
					return hasRows;
				}
			}
			catch (Exception ex)
			{
				SetLastError(ex.Message);
				return false;
			}
		}

		public string GetLastError()
		{
			return m_sLastError ?? string.Empty;
		}

		public InternalString QueryFirstRowSingleColumn(string sxSql)
		{
			if (m_pConn == null || sxSql == null)
			{
				ClearLastError();
				return null;
			}

			try
			{
				using (var cmd = new NpgsqlCommand(sxSql, m_pConn))
				using (var reader = cmd.ExecuteReader())
				{
					if (!reader.Read() || reader.FieldCount < 1)
						return null;

					object value = reader.GetValue(0);
					string text = value == null ? string.Empty : Convert.ToString(value);
					ClearLastError();
					return new InternalString(text);
				}
			}
			catch (Exception ex)
			{
				SetLastError(ex.Message);
				return null;
			}
		}

		private static List<string> SplitParamString(string sxParamString, int nParams)
		{
			var result = new List<string>(nParams > 0 ? nParams : 0);
			if (sxParamString == null || nParams <= 0)
				return result;

			int start = 0;
			for (int i = 0; i < nParams; i++)
			{
				if (start >= sxParamString.Length)
				{
					result.Add(string.Empty);
					continue;
				}

				int idx = sxParamString.IndexOf('\x01', start);
				if (idx < 0)
				{
					result.Add(sxParamString.Substring(start));
					start = sxParamString.Length;
				}
				else
				{
					result.Add(sxParamString.Substring(start, idx - start));
					start = idx + 1;
				}
			}

			return result;
		}

		private static string RewriteSqlForNpgsql(string sxSql, int nParams)
		{
			// Original SQL uses $1, $2 ... style. Npgsql by default expects @p1, @p2 ...
			// To avoid touching the NLL-side replacement logic, just adapt it here.
			if (sxSql == null || nParams <= 0)
				return sxSql;

			string sqlOut = sxSql;
			for (int i = 1; i <= nParams; i++)
			{
				string from = "$" + i.ToString();
				string to = "@p" + i.ToString();
				sqlOut = sqlOut.Replace(from, to);
			}
			return sqlOut;
		}

		public PgSqlResult ExecuteQuery(string sxSql, int nParams, string sxParamString)
		{
			if (m_pConn == null || sxSql == null)
			{
				ClearLastError();
				return null;
			}

			try
			{
				List<string> paramValues = SplitParamString(sxParamString, nParams);
				string finalSql = RewriteSqlForNpgsql(sxSql, nParams);

				using (var cmd = new NpgsqlCommand(finalSql, m_pConn))
				{
					for (int i = 0; i < paramValues.Count; i++)
					{
						string paramName = "p" + (i + 1).ToString();
						cmd.Parameters.AddWithValue(paramName, (object)paramValues[i] ?? DBNull.Value);
					}

					using (var reader = cmd.ExecuteReader())
					{
						var result = new PgSqlResult(reader);
						ClearLastError();
						return result;
					}
				}
			}
			catch (Exception ex)
			{
				SetLastError(ex.Message);
				return null;
			}
		}
	}
}

